### ISSUE TITLE
dart too slow

### ISSUE THREAD

**Issue by anpark** [dmlc/xgboost#1400]
i use dart for train, but it's too slow, time used about ten times more than base gbtree

**Comment by khotilov** [comment]
With tree dropout, the existing prediction buffering mechanism cannot be used. So, the more iterations you run, the slower becomes prediction.

**Comment by marugari** [comment]
We might be able to enable buffer in `Pred` and update buffer in `CommitModel`.

**Comment by anpark** [comment]
so we should only optimize predict function? 
@khotilov, just as you say, when run using 1000 rounds, the more iterations the slower. when about 300 rounds, it costs 3min each round and i have to stop the train

**Comment by anpark** [comment]
@marugari can you give some advice? thanks.

**Comment by marugari** [comment]
DART booster does not support buffer due to change of dropped trees' leaf scores, so booster must follow the path of all existing trees even though dropped trees are relatively few.

If we could use the existing prediction buffering mechanism in `Pred` and update buffer with change of leaf scores in `CommitModel`, DART booster could skip following the path of trees and be speeded up.
